### PR TITLE
add sequential pairwise distance constraint to ba

### PR DESCRIPTION
**Pull request by homee-james**
**Context:**

argument `fix_coord_system` in `colmap/estimators/bundle_adjusment.cc` only triggers distance constraint to first two cameras, which is not robust enough. 

**Changes**

add a new argument `fix_global_coord_system` to `colmap/estimators/bundle_adjusment.cc` to trigger distance constraint  to every image pair sequentially.

### CODE REVIEW

**Comment by homee-james** [comment]
in [7d2ec2c](https://github.com/homee-ai/colmap/pull/5/commits/7d2ec2cb69ef2c603f555f86b6393b95d332ae49)

1. change pairwise distance to pairwise translation vector
2. add rotation constraint
3. add weights for translation and rotation, respectively
4. use camera-to-world matrix to calculate pairwise translation and rotation

File: src/colmap/estimators/bundle_adjustment.h
Comment (homee-dennis) [inline src/colmap/estimators/bundle_adjustment.h:1]: Suggest adding .normalized() after quaternion multiplication

File: src/colmap/estimators/bundle_adjustment.cc [hunk 1]
```diff
@@ -462,6 +524,81 @@ void BundleAdjuster::AddCoordinateSystemConstraint(Reconstruction* reconstructio
...
+
+    // Calculate initial camera centers (in world coordinates)
+    const Eigen::Vector3d center1 = world_from_cam1.translation;
+    const Eigen::Vector3d center2 = world_from_cam2.translation;
+
+    // Calculate initial relative transformation between cameras
+    // relative_rotation = R2^(-1) * R1 where R1, R2 are world-from-camera rotations
+    const Eigen::Vector3d initial_translation_diff = center2 - center1;
+    const Eigen::Quaterniond initial_relative_rotation = 
+        image2.CamFromWorld().rotation * image1.CamFromWorld().rotation.inverse();
```
Comment (homee-dennis) [inline src/colmap/estimators/bundle_adjustment.cc]: Suggest adding .normalized() after quaternion multiplication

File: src/colmap/estimators/bundle_adjustment.h [hunk 2]
```diff
@@ -243,6 +255,59 @@ class BundleAdjuster {
...
+
+        // Convert target relative rotation to template type
+        const Eigen::Quaternion<T> target_relative_rotation(
+            T(relative_rotation_.w()),
+            T(relative_rotation_.x()),
+            T(relative_rotation_.y()),
+            T(relative_rotation_.z()));
+
+        // Compute the current relative rotation
+        const Eigen::Quaternion<T> current_relative_rotation = q2 * q1.inverse();
```
Comment (homee-dennis) [inline src/colmap/estimators/bundle_adjustment.h]: Suggest adding .normalized() after quaternion multiplication

File: src/colmap/estimators/bundle_adjustment.h [hunk 3]
```diff
@@ -243,6 +255,59 @@ class BundleAdjuster {
...
+
+        // Compute the error quaternion between target and current relative rotations
+        const Eigen::Quaternion<T> error = target_relative_rotation.inverse() * current_relative_rotation;
```
Comment (homee-dennis) [inline src/colmap/estimators/bundle_adjustment.h]: Suggest adding .normalized() after quaternion multiplication

**Comment by homee-dennis** [comment]
LGTM, just suggest adding .normalized() after quaternion multiplication

**Pull request merged**